pkg/kubemark/controller.go

```diff
@@ -0,0 +1,353 @@
+// provider for kubemark that would care for deleting the nodes.
+type kubemarkCluster struct {
+	client            kubeclient.Interface
+	nodeLister        listersv1.NodeLister
+	nodeListerSynced  cache.InformerSynced
+	nodesToDelete     map[string]bool
+	nodesToDeleteLock sync.Mutex
+}
+
+// NewController creates Controller using the provided clients to talk to external
+// and kubemark clusters.
+func NewController(externalClient kubeclient.Interface, externalInformerFactory informers.SharedInformerFactory,
+	kubemarkClient kubeclient.Interface, kubemarkNodeInformer informersv1.NodeInformer) (*Controller, error) {
+	controller := &Controller{
+		externalCluster: externalCluster{
+			rcLister:  externalInformerFactory.InformerFor(&apiv1.ReplicationController{}, newReplicationControllerInformer),
+			podLister: externalInformerFactory.InformerFor(&apiv1.Pod{}, newPodInformer),
+			client:    externalClient,
+		},
+		kubemarkCluster: kubemarkCluster{
+			nodeLister:        kubemarkNodeInformer.Lister(),
+			nodeListerSynced:  kubemarkNodeInformer.Informer().HasSynced,
+			client:            kubemarkClient,
+			nodesToDelete:     make(map[string]bool),
+			nodesToDeleteLock: sync.Mutex{},
+		},
+	}
+
+	controller.externalCluster.rcListerSynced = controller.externalCluster.rcLister.HasSynced
+	controller.externalCluster.podListerSynced = controller.externalCluster.podLister.HasSynced
+
+	kubemarkNodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: controller.kubemarkCluster.removeUnneededNodes,
+	})
+
+	rand.Seed(time.Now().UTC().UnixNano())
+
+	return controller, nil
+}
+
+// Run waits for population of caches and populates the node template needed
+// for creation of kubemark nodes.
+func (kubemarkController *Controller) Run(stopCh chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	glog.Infof("starting kubemark controller")
+	defer glog.Infof("shutting down kubemark controller")
+
+	if !controller.WaitForCacheSync("kubemark", stopCh,
+		kubemarkController.externalCluster.rcListerSynced,
+		kubemarkController.externalCluster.podListerSynced,
+		kubemarkController.kubemarkCluster.nodeListerSynced) {
+		return
+	}
+
+	// Get hollow node template from an existing hollow node to be able to create
+	// new nodes based on it.
+	nodeTemplate, err := kubemarkController.getNodeTemplate()
+	if err != nil {
+		glog.Fatalf("Failed to get node template: %s", err)
+	}
+	kubemarkController.nodeTemplate = nodeTemplate
+
+	<-stopCh
+}
+
+// GetNodesForNodegroup returns list of the nodes in the node group.
+func (kubemarkController *Controller) GetNodesForNodegroup(nodeGroup string) ([]string, error) {
+	pods := kubemarkController.externalCluster.podLister.GetStore().List()
+	result := make([]string, 0, len(pods))
+	for _, podObj := range pods {
+		pod, ok := podObj.(*apiv1.Pod)
+		if !ok {
+			continue
+		}
+		if pod.ObjectMeta.Labels[nodeGroupLabel] == nodeGroup {
```

Comment:
No need to List() all the nodes from the store and then manually filter on a given label. You can directly do `podLister.List(<label-selector>)`
For e.g - https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podgc/gc_controller_test.go#L203